modules/references/server/controllers/reference.server.controller.js

```diff
@@ -345,3 +350,74 @@ exports.readMany = function readMany(req, res, next) {
 
   ], processResponses.bind(this, res, next));
 };
+
+
+/**
+ * Validator for id of referenceById controller
+ * @param {string} id - referenceId
+ */
+function validateReadOne(id) {
+  var valid = true;
+  var details = {};
+
+  var isIdValid = mongoose.Types.ObjectId.isValid(id);
+  if (!isIdValid) {
+    valid = false;
+    details.referenceId = 'invalid';
+  }
+
+  return { valid: valid, details: details };
+}
+
+/**
+ * Load a reference by id to request.reference
+ */
+exports.referenceById = function referenceById(req, res, next, id) { // eslint-disable-line no-unused-vars
+  // don't bother fetching a reference for non-public users or guests
+  if (!req.user || !req.user.public) return next();
+
+  async.waterfall([
+    // validate
+    validate.bind(this, validateReadOne, id),
+    // find the reference by id
+    function (cb) {
+      Reference.findById(req.params.referenceId)
+        .select(referenceFields)
+        .populate('userFrom userTo', userProfile.userMiniProfileFields)
+        .exec(cb);
+    },
+    // make sure that nonpublic references are not exposed
+    // assign reference to request object
+    function (reference, cb) {
+
+      // nonpublic reference can be exposed to userFrom or userTo only.
+      var isExistentPublicOrFromToSelf = reference
+        && (reference.public || [reference.userFrom._id.toString(), reference.userTo._id.toString()].includes(req.user._id.toString()));
+      if (!isExistentPublicOrFromToSelf) {
+        return cb({
+          status: 404,
+          body: {
+            errType: 'not-found',
+            details: {
+              reference: 'not found'
+            }
+          }
+        });
+      }
+
+      // assign the reference to the request object
+      // when reference is public, only userFrom can see it whole
+      var isUserFrom = reference.userFrom._id.toString() === req.user._id.toString();
```

Comment:
I believe you can compare Mongo ObjectIds with `userFrom._id.equals(req.user._id)`

Reply:
This and other similar comparisons fixed in a4f41ec.